### PR TITLE
260 boost constraint

### LINKED ISSUE #260: Support BOOST constraint
Needed by Nova tutorial 3.1, among others.

### PR DESCRIPTION
**Pull request by aminere**
Testable with puzzle `9386704` on eterna-dev
resolves #260

### CODE REVIEW

**Comment by aminere** [comment]
@luxaritas No prob for 1 and 2,
About the BoostLoops class, I didn't make it as a pure function because it needs state (see the `allLoops` member). In Nova this state is stored within the puzzle. Apart from this, it's just ported from actionscript so no need to waste time reviewing its count() and detectLoops() methods imo.

**Comment by luxaritas** [comment]
It looks to me like detectLoops is just generating allLoops (it's essentially acting as a cache). Should be easy enough to have detectLoops create an array and return it as opposed to adding it to allLoops, and have count generate it each time.

**Comment by aminere** [comment]
@luxaritas yes that would work. Those loops are cached in Nova so I kept the same behavior. I don't think it's a problem to not cache and generate them everytime. Will do!

File: src/eterna/constraints/Constraint.ts [hunk 1]
```diff
@@ -15,6 +15,15 @@ export interface HighlightInfo {
     stateIndex?: number;
 }
 
+export interface ConstraintContext {
+    undoBlocks: UndoBlock[];
+    targetConditions?: any[];
+    puzzle?: Puzzle;
+    targetPairs?: number[][];
+    currentTargetIndex?: number;
+    sequence?: number[];
```
Comment (luxaritas) [inline src/eterna/constraints/Constraint.ts:24]: One last thing to comment on here. targetPairs and sequence are already on undoBlock, so you don't need to pass them separately. Also, currentTargetIndex should not have an impact on how the constraint is evaluated - if a constraint doesn't have a parameter specifying the state, that means currently we only support it in state 1 (ie, index 0) so you can hardcode that in for the boost constraint.

Comment (aminere) [reply src/eterna/constraints/Constraint.ts:24]: OK will do. Note that Nova pulls both `targetPairs` and `sequence` from PoseEdit state, from what you're saying, I assume `undoBlock[0]` is pointing to the same thing.

Comment (aminere) [reply src/eterna/constraints/Constraint.ts]: Done

**Pull request merged**